lib/winapi/sync.c

```diff
@@ -1,11 +1,145 @@
 #include <synchapi.h>
 #include <assert.h>
+#include <stdatomic.h>
 #include <stdbool.h>
 #include <processthreadsapi.h>
 #include <winbase.h>
 #include <winerror.h>
 #include <xboxkrnl/xboxkrnl.h>
 
+static BOOL WINAPI condvar_init (PINIT_ONCE InitOnce, PVOID Parameter, PVOID *Context)
+{
+    PCONDITION_VARIABLE ConditionVariable = Parameter;
+    NTSTATUS status;
+    status = NtCreateEvent(&ConditionVariable->eventHandles[0], NULL, SynchronizationEvent, FALSE);
+    assert(NT_SUCCESS(status));
+    status = NtCreateEvent(&ConditionVariable->eventHandles[1], NULL, NotificationEvent, FALSE);
+    assert(NT_SUCCESS(status));
+    return true;
+}
+
+VOID InitializeConditionVariable (PCONDITION_VARIABLE ConditionVariable)
+{
+    ConditionVariable->initOnce.Ptr = 0;
+    ConditionVariable->eventHandles[0] = INVALID_HANDLE_VALUE;
+    ConditionVariable->eventHandles[1] = INVALID_HANDLE_VALUE;
+    atomic_store(&ConditionVariable->waitCount, 0);
+
+    InitOnceExecuteOnce(&ConditionVariable->initOnce, condvar_init, ConditionVariable, NULL);
+}
+
+BOOL SleepConditionVariableCS (PCONDITION_VARIABLE ConditionVariable, PCRITICAL_SECTION CriticalSection, DWORD dwMilliseconds)
+{
+    InitOnceExecuteOnce(&ConditionVariable->initOnce, condvar_init, ConditionVariable, NULL);
+
+    NTSTATUS status;
+    LARGE_INTEGER waitTime;
+    LARGE_INTEGER *waitTimePtr = NULL;
+
+    if (dwMilliseconds != INFINITE) {
+        waitTime.QuadPart = dwMilliseconds * 1000;
+        waitTimePtr = &waitTime;
+    }
+
+    LeaveCriticalSection(CriticalSection);
+    atomic_fetch_add(&ConditionVariable->waitCount, 1);
+    status = NtWaitForMultipleObjectsEx(2, ConditionVariable->eventHandles, WaitAny, UserMode, FALSE, waitTimePtr);
+    if (status == WAIT_FAILED || status == STATUS_TIMEOUT) {
+        atomic_fetch_sub(&ConditionVariable->waitCount, 1);
+        EnterCriticalSection(CriticalSection);
+        return FALSE;
+    }
+
+    int oldCount = atomic_fetch_sub(&ConditionVariable->waitCount, 1);
+    if (oldCount == 1) {
+        // Last waiter needs to manually reset the broadcast-object
+        status = NtClearEvent(ConditionVariable->eventHandles[1]);
+        if (!NT_SUCCESS(status)) {
+            EnterCriticalSection(CriticalSection);
+            return FALSE;
+        }
+    }
+
+    EnterCriticalSection(CriticalSection);
+    return TRUE;
+}
+
+BOOL SleepConditionVariableSRW (PCONDITION_VARIABLE ConditionVariable, PSRWLOCK SRWLock, DWORD dwMilliseconds, ULONG Flags)
```

Comment:
nit: There's a lot of redundancy with `SleepConditionVariableCS` that could have been factored out; fine as-is though.